### PR TITLE
Realtime Collaboration System

### PR DESCRIPTION
**Pull request by Vedant1202**
This PR aims to develop a realtime collaboration system to enable users to work together in real time.

A demo link for features currently implemented is available here -> [Camicroscope Real Time Collaboration - v1.0](https://youtu.be/jT4tpCJrILU)

This PR depends on [Distro PR-173](https://github.com/camicroscope/Distro/pull/173)
and [caMicroscope PR-531](https://github.com/camicroscope/caMicroscope/pull/531)

## Summary
Features included:
- In-viewer collaboration (currently limited to annotations/draw tool), using async request interceptors.
- Manage collaboration room in `table.html` and `viewer.html`. 
- Granting and revoking access to rooms
- Role based actions
- Securing sockets connections using JWT
- Video/Voice call using Jitsi Meet

## Motivation
Since the pandemic, most people have switched to working remotely, and this feature would enable them to work remotely.

## Testing
Tested on Windows 10 Home, using 
- Docker Desktop, running on Docker Engine v20.10.6
- Microsoft Edge Version 91.0.864.59 (Official build) (64-bit)

### Features yet to implement ->
- Chat window
- Extending collaborations to other features such as heatmaps, etc
- Extending Role based actions
- Cleaning up code and removing comments

### Problems yet to be solved
- Cross client socket communication not working on docker, but is working when setup is done outside of docker.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "test": "mocha test/* --exit",
     "lint": "eslint *.js handlers/*.js test/**/*.js",
     "lint-fix": "eslint *.js handlers/*.js test/**/*.js --fix",
-    "start": "node caracal.js"
+    "start": "nodemon caracal.js"
```
Comment (YashKumarVerma) [inline package.json:10]: i think the start command should be via node and not nodemon.

Comment (YashKumarVerma) [reply package.json:10]: we could keep a separate command like `start:dev` or something similar to have hot reload

**Review by YashKumarVerma — COMMENTED** [review]
i'll try running it by end of day

File: package.json (hunk 1 above)
Comment (Vedant1202) [reply package.json:10]: Yes that's a good point

File: Dockerfile [hunk 2]
```diff
@@ -6,12 +6,16 @@ WORKDIR /src
 RUN npm install
 ARG viewer
 ARG fork
-RUN git clone https://github.com/${fork:-camicroscope}/camicroscope.git --branch=${viewer:-master}
+# RUN git clone https://github.com/${fork:-camicroscope}/camicroscope.git --branch=${viewer:-master}
```
Comment (akhil-rana) [inline Dockerfile:9]: @Vedant1202 just remember to un-comment it again before we merge :)

File: static/login.html [hunk 3]
```diff
@@ -3,7 +3,7 @@
 <head>
   <title>CaMicroscope</title>
   <meta charset="utf-8" />
-  <meta name="google-signin-client_id" content="539699277901-buh78f048405grejogqbrskj233rt0d0.apps.googleusercontent.com">
+  <meta name="google-signin-client_id" content="208463168480-m5fronlc3ic98elbicq7gmglst8trdkg.apps.googleusercontent.com">
```
Comment (akhil-rana) [inline static/login.html:6]: @Vedant1202 we should probably reset this to previous value.

**Review by birm — APPROVED** [review]
Approved pending @akhil-rana's comments on restoring some values/comments

File: Dockerfile (hunk 2 above)
Comment (Vedant1202) [reply Dockerfile]: done

**Comment by Vedant1202** [comment]
@birm done with restoring values

**Pull request merged**